pkg/rke2/failure_domain.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rke2
+
+import (
+	"sort"
+
+	"k8s.io/klog/v2/klogr"
+	"k8s.io/utils/pointer"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/collections"
+)
+
+// Log is the global logger for the internal package.
+var Log = klogr.New()
+
+type failureDomainAggregation struct {
+	id    string
+	count int
+}
+type failureDomainAggregations []failureDomainAggregation
+
+// Len is the number of elements in the collection.
+func (f failureDomainAggregations) Len() int {
+	return len(f)
+}
+
+// Less reports whether the element with
+// index i should sort before the element with index j.
+func (f failureDomainAggregations) Less(i, j int) bool {
+	return f[i].count < f[j].count
+}
+
+// Swap swaps the elements with indexes i and j.
+func (f failureDomainAggregations) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+
+// PickMost returns a failure domain that is in machines and has most control-plane machines on.
+func PickMost(c *ControlPlane, machines collections.Machines) *string {
```

Comment:
Whats the reason for not using the upstream version? https://github.com/kubernetes-sigs/cluster-api/blob/9c5fdae1cf23b3a088181363b5aa7e34c537f0b0/util/failuredomains/failure_domains.go#L53